Compile error on modern g++

Hi, I've just see this under linux g++10

```
[ 57%] Building CXX object CMakeFiles/igel.dir/src/search.cpp.o
/ssd/engines/igel/src/search.cpp: In member function ‘void Search::clearStacks()’:
/ssd/engines/igel/src/search.cpp:652:47: error: ‘void* memset(void*, int, size_t)’ clearing an object of non-trivial type ‘class Move’; use assignment or value-initialization instead [-Werror=class-memaccess]
  652 |     memset(m_moveStack, 0, sizeof(m_moveStack));
      |                                               ^
In file included from /ssd/engines/igel/src/moves.h:24,
                 from /ssd/engines/igel/src/search.cpp:22:
/ssd/engines/igel/src/position.h:43:7: note: ‘class Move’ declared here
   43 | class Move
      |       ^~~~
/ssd/engines/igel/src/search.cpp:660:69: error: ‘void* memset(void*, int, size_t)’ clearing an object of non-trivial type ‘class Move’; use assignment or value-initialization instead [-Werror=class-memaccess]
  660 |         memset(m_threadParams[i].m_moveStack, 0, sizeof(m_moveStack));
      |                                                                     ^
In file included from /ssd/engines/igel/src/moves.h:24,
                 from /ssd/engines/igel/src/search.cpp:22:
/ssd/engines/igel/src/position.h:43:7: note: ‘class Move’ declared here
   43 | class Move
```

--- a/src/misc.cpp
+++ b/src/misc.cpp
@@ -21,10 +21,19 @@
 #include <malloc.h>
 #include <cstdlib>
 
+// trick from Stockfish to fix aligned allocation issues on Android
+#if (defined(__APPLE__) && defined(_LIBCPP_HAS_C11_FEATURES)) || defined(__ANDROID__) || defined(__OpenBSD__) || (defined(__GLIBCXX__) && !defined(_GLIBCXX_HAVE_ALIGNED_ALLOC) && !defined(_WIN32))
+#define POSIXALIGNEDALLOC
+#include <stdlib.h>
+#endif
+
 void* std_aligned_alloc(size_t alignment, size_t size) {
-#if defined(__APPLE__)
-    return aligned_alloc(alignment, size);
-#elif defined(_WIN32)
+#if defined(POSIXALIGNEDALLOC)
+    void* pointer;
+    if (posix_memalign(&pointer, alignment, size) == 0)
+        return pointer;
+    return nullptr;
+#elif (defined(_WIN32) || (defined(__APPLE__) && !defined(_LIBCPP_HAS_C11_FEATURES)))
     return _mm_malloc(size, alignment);
 #else
     return std::aligned_alloc(alignment, size);
@@ -32,12 +41,10 @@ void* std_aligned_alloc(size_t alignment, size_t size) {
 }
 
 void std_aligned_free(void* ptr) {
-#if defined(__APPLE__)
+#if defined(POSIXALIGNEDALLOC)
     free(ptr);
-#elif defined(_WIN32)
+#elif (defined(_WIN32) || (defined(__APPLE__) && !defined(_LIBCPP_HAS_C11_FEATURES)))
     _mm_free(ptr);
-#else
-    free(ptr);
 #endif
 }
 #endif

--- a/src/position.cpp
+++ b/src/position.cpp
@@ -64,9 +64,12 @@ void Position::Clear()
     std::memset(evalList.piece_id_list, 0, sizeof(evalList.piece_id_list));
     std::memset(evalList.pieceListFw, 0, sizeof(evalList.pieceListFw));
     std::memset(evalList.pieceListFb, 0, sizeof(evalList.pieceListFb));
-    std::memset(m_undos, 0, sizeof(m_undos));
+
+    for (unsigned int i = 0; i < sizeof(m_undos) / sizeof(Undo); ++i)
+        m_undos->reset();
 
     m_state = &m_undos[0];
+
 #endif
 
     for (FLD f = 0; f < 64; ++f)

--- a/src/position.h
+++ b/src/position.h
@@ -58,7 +58,7 @@ public:
     FLD Piece() const { return (m_data >> 20) & 0x0f; }
     FLD Captured() const { return (m_data >> 16) & 0x0f; }
     FLD Promotion() const { return (m_data >> 12) & 0x0f; }
-
+    inline void reset() { m_data = 0; }
     operator U32() const { return m_data; }
 
 private:
@@ -95,6 +95,38 @@ struct Undo
     DirtyPiece dirtyPiece;
     Undo* previous;
 #endif
+
+    void reset() 
+    {
+        m_castlings = 0;
+        m_ep        = 0;
+        m_fifty     = 0;
+        m_hash      = 0;
+
+        m_mv.reset();
+
+#if defined(EVAL_NNUE)
+        std::memset(&accumulator.accumulation, 0, sizeof(accumulator.accumulation));
+
+        accumulator.score                   = VALUE_ZERO;
+        accumulator.computed_accumulation   = false;
+        accumulator.computed_score          = false;
+
+        dirtyPiece.dirty_num  = 0;
+        dirtyPiece.pieceId[0] = PIECE_ID_ZERO;
+        dirtyPiece.pieceId[1] = PIECE_ID_ZERO;
+
+        for (unsigned int i = 0; i < sizeof(dirtyPiece.old_piece) / sizeof(ExtPieceSquare); ++i)
+            for (unsigned int j = 0; j < sizeof(dirtyPiece.old_piece[i].from) / sizeof(PieceSquare); ++j)
+                dirtyPiece.old_piece[i].from[j] = PS_NONE;
+
+        for (unsigned int i = 0; i < sizeof(dirtyPiece.new_piece) / sizeof(ExtPieceSquare); ++i)
+            for (unsigned int j = 0; j < sizeof(dirtyPiece.new_piece[i].from) / sizeof(PieceSquare); ++j)
+                dirtyPiece.new_piece[i].from[j] = PS_NONE;
+
+        previous = nullptr;
+#endif
+    }
 };
 
 class Position

--- a/src/search.cpp
+++ b/src/search.cpp
@@ -657,7 +657,9 @@ void Search::clearKillers()
 
 void Search::clearStacks()
 {
-    memset(m_moveStack, 0, sizeof(m_moveStack));
+    for (unsigned int j = 0; j < sizeof(m_moveStack) / sizeof(Move); ++j)
+        m_moveStack[j].reset();
+
     memset(m_pieceStack, 0, sizeof(m_pieceStack));
     memset(m_followTable, 0, sizeof(m_followTable));
     memset(m_counterTable, 0, sizeof(m_counterTable));
@@ -665,7 +667,10 @@ void Search::clearStacks()
     memset(m_pvSize, 0, sizeof(m_pvSize));
 
     for (unsigned int i = 0; i < m_thc; ++i) {
-        memset(m_threadParams[i].m_moveStack, 0, sizeof(m_moveStack));
+
+        for (unsigned int j = 0; j < sizeof(m_threadParams[i].m_moveStack) / sizeof(Move); ++j)
+            m_threadParams[i].m_moveStack[j].reset();
+
         memset(m_threadParams[i].m_pieceStack, 0, sizeof(m_pieceStack));
         memset(m_threadParams[i].m_followTable, 0, sizeof(m_followTable));
         memset(m_threadParams[i].m_counterTable, 0, sizeof(m_counterTable));

--- a/src/uci.cpp
+++ b/src/uci.cpp
@@ -30,7 +30,7 @@
 #include <iostream>
 #include <sstream>
 
-const std::string VERSION = "3.0-dev-14";
+const std::string VERSION = "3.0-dev-15";
 
 #if defined(ENV64BIT)
     #if defined(_BTYPE)

